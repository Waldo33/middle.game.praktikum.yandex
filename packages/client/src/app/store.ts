import { configureStore } from '@reduxjs/toolkit'
import authReducer from '../processes/auth/model/authSlice'

export interface Signin {
  login: string
  password: string
}

export interface User {
  id: number
  first_name: string
  second_name: string
  display_name: string
  login: string
  email: string
  password: string
  phone: string
  avatar: string
  reason?: string
}

export interface Auth {
  isAuthenticated: undefined | boolean
  user: null | User
  error: null | string
  loading: boolean
}
export interface Store {
  auth: Auth
}

const store = configureStore({
  reducer: {
    auth: authReducer,
  },
})

export default store
