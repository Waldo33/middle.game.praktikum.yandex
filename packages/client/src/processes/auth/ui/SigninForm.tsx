import { FC } from 'react'
import { useNavigate } from 'react-router-dom'

import { useDispatch, useSelector } from 'react-redux'
import { signinThunk } from '../model/thunks'
import { selectAuthLoading } from '../model/selectors'

import { zodResolver } from '@hookform/resolvers/zod'
import { useForm } from 'react-hook-form'
import { z } from 'zod'
import { Button } from '@shared/components/ui/button'
import { Form } from '@shared/components/ui/form'

import { validationRules } from '@shared/config/validationRules'

import { AppDispatch } from '@app/store'
import { ROUTES } from '@shared/config/routes'
import { useToast } from '@shared/hooks/use-toast'
import { FormFieldWrapper } from './FormFieldWrapper'

const formSchema = z.object({
  login: validationRules.login,
  password: validationRules.password,
})

const SigninForm: FC = () => {
  const dispatch: AppDispatch = useDispatch()
  const navigate = useNavigate()
  const { toast } = useToast()

  const loading = useSelector(selectAuthLoading)

  const form = useForm<z.infer<typeof formSchema>>({
    resolver: zodResolver(formSchema),
    defaultValues: {
      login: '',
      password: '',
    },
  })

  const onSubmit = async (values: z.infer<typeof formSchema>) => {
    const resultAction = await dispatch(signinThunk(values))

    if (resultAction.payload === true) {
      navigate(ROUTES.GAME)
    } else {
      toast({
        description: resultAction.payload,
        variant: 'destructive',
      })
    }
  }

  return (
    <Form {...form}>
      <form onSubmit={form.handleSubmit(onSubmit)} className="space-y-8">
        <FormFieldWrapper control={form.control} name="login" label="Login" />
        <FormFieldWrapper
          control={form.control}
          name="password"
          label="Password"
        />
        <Button className="w-full" type="submit">
          {loading ? 'Logging in...' : 'Submit'}
        </Button>
      </form>
    </Form>
  )
}

export default SigninForm
