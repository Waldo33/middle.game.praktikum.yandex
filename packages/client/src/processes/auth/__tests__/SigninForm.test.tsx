import React from 'react'
import { render } from '@testing-library/react'
import { Provider } from 'react-redux'
import store from '@app/store'
import SigninForm from '../ui/SigninForm'
import { BrowserRouter as Router } from 'react-router-dom'

jest.mock('@processes/auth/api/authApi', () => ({
  BASE_AUTH_API: 'https://ya-praktikum.tech/api/v2',
  signinApi: jest.fn(),
}))

describe('SigninForm', () => {
  it('соответствует снапшоту до auth/loading', () => {
    const { asFragment } = render(
      <Provider store={store}>
        <Router>
          <SigninForm />
        </Router>
      </Provider>
    )

    expect(asFragment()).toMatchSnapshot()
  })

  it('соответствует снапшоту после auth/loading', () => {
    const { asFragment } = render(
      <Provider store={store}>
        <Router>
          <SigninForm />
        </Router>
      </Provider>
    )

    store.dispatch({ type: 'auth/loading' })
    expect(asFragment()).toMatchSnapshot()
  })
})
