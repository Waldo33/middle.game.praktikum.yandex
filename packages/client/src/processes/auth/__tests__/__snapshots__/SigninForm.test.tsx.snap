// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SigninForm соответствует снапшоту до auth/loading 1`] = `
<DocumentFragment>
  <form
    class="space-y-8"
  >
    <div
      class="space-y-2"
    >
      <label
        class="text-sm font-medium leading-none peer-disabled:cursor-not-allowed peer-disabled:opacity-70"
        for=":r0:-form-item"
      >
        Login
      </label>
      <input
        aria-describedby=":r0:-form-item-description"
        aria-invalid="false"
        class="flex h-10 w-full rounded-md border border-input bg-background px-3 py-2 text-sm ring-offset-background file:border-0 file:bg-transparent file:text-sm file:font-medium file:text-foreground placeholder:text-muted-foreground focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:cursor-not-allowed disabled:opacity-50"
        id=":r0:-form-item"
        name="login"
        value=""
      />
    </div>
    <div
      class="space-y-2"
    >
      <label
        class="text-sm font-medium leading-none peer-disabled:cursor-not-allowed peer-disabled:opacity-70"
        for=":r1:-form-item"
      >
        Password
      </label>
      <input
        aria-describedby=":r1:-form-item-description"
        aria-invalid="false"
        class="flex h-10 w-full rounded-md border border-input bg-background px-3 py-2 text-sm ring-offset-background file:border-0 file:bg-transparent file:text-sm file:font-medium file:text-foreground placeholder:text-muted-foreground focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:cursor-not-allowed disabled:opacity-50"
        id=":r1:-form-item"
        name="password"
        value=""
      />
    </div>
    <button
      class="inline-flex items-center justify-center gap-2 whitespace-nowrap rounded-md text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg]:size-4 [&_svg]:shrink-0 bg-primary text-primary-foreground hover:bg-primary/90 h-10 px-4 py-2 w-full"
      type="submit"
    >
      Submit
    </button>
  </form>
</DocumentFragment>
`;

exports[`SigninForm соответствует снапшоту после auth/loading 1`] = `
<DocumentFragment>
  <form
    class="space-y-8"
  >
    <div
      class="space-y-2"
    >
      <label
        class="text-sm font-medium leading-none peer-disabled:cursor-not-allowed peer-disabled:opacity-70"
        for=":r2:-form-item"
      >
        Login
      </label>
      <input
        aria-describedby=":r2:-form-item-description"
        aria-invalid="false"
        class="flex h-10 w-full rounded-md border border-input bg-background px-3 py-2 text-sm ring-offset-background file:border-0 file:bg-transparent file:text-sm file:font-medium file:text-foreground placeholder:text-muted-foreground focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:cursor-not-allowed disabled:opacity-50"
        id=":r2:-form-item"
        name="login"
        value=""
      />
    </div>
    <div
      class="space-y-2"
    >
      <label
        class="text-sm font-medium leading-none peer-disabled:cursor-not-allowed peer-disabled:opacity-70"
        for=":r3:-form-item"
      >
        Password
      </label>
      <input
        aria-describedby=":r3:-form-item-description"
        aria-invalid="false"
        class="flex h-10 w-full rounded-md border border-input bg-background px-3 py-2 text-sm ring-offset-background file:border-0 file:bg-transparent file:text-sm file:font-medium file:text-foreground placeholder:text-muted-foreground focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:cursor-not-allowed disabled:opacity-50"
        id=":r3:-form-item"
        name="password"
        value=""
      />
    </div>
    <button
      class="inline-flex items-center justify-center gap-2 whitespace-nowrap rounded-md text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg]:size-4 [&_svg]:shrink-0 bg-primary text-primary-foreground hover:bg-primary/90 h-10 px-4 py-2 w-full"
      type="submit"
    >
      Submit
    </button>
  </form>
</DocumentFragment>
`;
